### PR TITLE
Parallelize Construct and Call calls in the Node.js SDK

### LINKED ISSUE #7629: Support parallel construct requests for nodejs multi-lang components
For stability and due to global state issues in the nodejs runtime, we serialized MLC construct requests in https://github.com/pulumi/pulumi/pull/6452

This has some obvious downsides like not creating cloud resources in parallel if you are instantiating multiple instances of the same component in a program. 

Will require some of the same set of work as https://github.com/pulumi/pulumi/issues/5449 to remove global state from the nodejs runtime.

### PR DESCRIPTION
**Pull request by flostadler**
# Description

Due to global state issues in the nodejs runtime, we serialized MLC construct requests in #6452.
This had the downside that cloud resources were not created in parallel if multiple instances of the same component were instantiated in the same program.

This change removes the serialization of Construct and Call calls and moves the uncaught error handling into a generic handler that keeps track of all inflight Call/Construct callbacks and responds to them in case of a uncaught error in user code.
The global state issues were already fixed in #10568

Fixes #7629 

## Checklist

- [x] I have run `make tidy` to update any new dependencies
- [x] I have run `make lint` to verify my code passes the lint check
  - [x] I have formatted my code using `gofumpt`

- [ ] I have added tests that prove my fix is effective or that my feature works

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

**Comment by flostadler** [comment]
I still need to figure out how to properly test this. There's existing tests covering this code (including the uncaught error handler), but I'd like to create a test that tests the parallel execution of MLC construct requests and Call requests.

**Comment by flostadler** [comment]
This line here mentions global state: https://github.com/pulumi/pulumi/blob/master/sdk/nodejs/provider/server.ts#L494-L495

I had a look at what's happening there and this seems to not be accurate anymore. `settings` and `config` have been changed in #10568 to use `AsyncLocalStorage`.

**Comment by flostadler** [comment]
> I still need to figure out how to properly test this. There's existing tests covering this code (including the uncaught error handler), but I'd like to create a test that tests the parallel execution of MLC construct requests and Call requests.

I found that this [test](https://github.com/pulumi/pulumi/blob/master/tests/integration/construct_component/nodejs/index.ts) is already creating multiple components, but only three.
Curious what y'all think about that.

File: sdk/nodejs/provider/server.ts [hunk 1]
```diff
@@ -37,14 +37,26 @@ class Server implements grpc.UntypedServiceImplementation {
     engineAddr: string | undefined;
     readonly provider: Provider;
     readonly uncaughtErrors: Set<Error>;
-
-    /** Queue of construct calls. */
-    constructCallQueue = Promise.resolve();
+    private readonly _callbacks = new Map<string, any>();
```
Comment (lunaris) [inline sdk/nodejs/provider/server.ts:40]: Can we do better than `any` for the type here? It seems like they are all at least `(err: Error, ...) => void`, maybe?

File: sdk/nodejs/provider/server.ts [hunk 2]
```diff
@@ -37,14 +37,26 @@ class Server implements grpc.UntypedServiceImplementation {
...
+        const uncaughtHandler = (err: Error) => {
+            if (!this.uncaughtErrors.has(err)) {
+                this.uncaughtErrors.add(err);
+            }
+            // terminate the outstanding gRPC requests.
+            this._callbacks.forEach((callback) => callback(err, undefined));
+        };
+        process.on("uncaughtException", uncaughtHandler);
+        // @ts-ignore 'unhandledRejection' will almost always invoke uncaughtHandler with an Error. so
+        // just suppress the TS strictness here.
```
Comment (lunaris) [inline sdk/nodejs/provider/server.ts:58]: What are the edge cases? Wondering for instance if it'd be better to give `uncaughtHandler` a union type or something and `if` that out with an error that'd be easier to debug if that happened?

File: sdk/nodejs/provider/server.ts [hunk 3]
```diff
@@ -272,33 +284,8 @@ class Server implements grpc.UntypedServiceImplementation {
...
-            }
-            // bubble the uncaught error in the user code back and terminate the outstanding gRPC request.
-            callback(err, undefined);
-        };
-        process.on("uncaughtException", uncaughtHandler);
-        // @ts-ignore 'unhandledRejection' will almost always invoke uncaughtHandler with an Error. so
-        // just suppress the TS strictness here.
-        process.on("unhandledRejection", uncaughtHandler);
+        const callbackId = Symbol("id");
+        this._callbacks.set(callbackId.toString(), callback);
```
Comment (lunaris) [inline sdk/nodejs/provider/server.ts:288]: I need to check but I feel like you could use the `Symbol` directly as a `Map` key, so `Map<Symbol, ...>`? But I could be wrong/there could be a good reason not to do this.

Comment (flostadler) [reply sdk/nodejs/provider/server.ts:288]: Ah you're right. Wasn't aware of that!

File: sdk/nodejs/provider/server.ts (hunk 2 above)
Comment (flostadler) [reply sdk/nodejs/provider/server.ts:58]: I replicated the existing error handler this replaces and this one here: 
https://github.com/pulumi/pulumi/blob/master/sdk/nodejs/provider/server.ts#L641-L654

This was around for quite some time already (4+ years) and I was hesitant to change the behavior not to introduce bugs. If we'd like to change those I'd propose to do it in a separate PR in order to isolate the changes

File: sdk/nodejs/provider/server.ts (hunk 1 above)
Comment (flostadler) [reply sdk/nodejs/provider/server.ts:40]: We could, but all the rpc methods are completely untyped: https://github.com/pulumi/pulumi/blame/master/sdk/nodejs/provider/server.ts#L381

Shouldn't we rather add types here first?

**Comment by flostadler** [comment]
I tried stringing together some initial unit tests for `provider/server.ts`, but I didn't manage to do so because under the hood it calls out to `configureRuntime` which ends up trying to make grpc calls to the monitor: https://github.com/pulumi/pulumi/blob/6a48d609d881422c45482d6e4027c4de6c3eb3e3/sdk/nodejs/provider/server.ts#L511

I feel like we'd need some code-restructuring to make this testable

**Review by lunaris — APPROVED** [review]

**Comment by justinvp** [comment]
Sorry for not asking about this before it was merged, but I would have thought we'd have to do something like this inside the implementation of `Construct` and `Call`?

https://github.com/pulumi/pulumi/blob/3ac40e019d10ae08de88585c372247872fe4b1a2/sdk/nodejs/automation/server.ts#L62-L64

**Comment by flostadler** [comment]
> pulumi/sdk/nodejs/automation/server.ts

Looking further into it I think you're right. Initially I was under the impression that this was handled in `settings` itself but it just uses a global store if running outside of `asyncLocalStorage.run`.
Thanks for pointing that out!

**Pull request merged**